### ISSUE TITLE
[accidental event] Websocket server can't  receive metadata before audio streaming starts.

### ISSUE THREAD

**Issue by chinarui-na** [amigniter/mod_audio_stream#44]
I want voice recognition on the called party in real time. I was called the Connect event again, initiating a two-way media bug.
caller
```
uuid_audio_stream channelId_caller start ws://127.0.0.1:3001 mono 8k caller_number
```
callee
```
uuid_audio_stream channelId_callee start ws://127.0.0.1:3001 mono 8k callee_number
```
websocket receives two connections, but only once metadata, sometimes caller_number, sometimes callee_number, or caller and callee.

my websocket handler code 
```
public class AudioStreamHandler extends SimpleChannelInboundHandler<WebSocketFrame> {
    private static final String FILE_NAME = "received_data.raw";
    private WebSocketClient webSocketClient;
    private HashMap<ChannelId,WebSocketClient> webSocketClientHashMap = new HashMap<>();
    private HashMap<ChannelId,byte []> bytesMap = new HashMap<>();
    public String decodeUTF8(byte[] utf8Bytes) {
        if (utf8Bytes == null) {
            return null;
        }
        return new String(utf8Bytes, StandardCharsets.UTF_8);
    }

    @Override
    protected void channelRead0(ChannelHandlerContext ctx, WebSocketFrame frame) throws Exception {
        if (frame instanceof TextWebSocketFrame) {
            TextWebSocketFrame msgContent = (TextWebSocketFrame) frame;
            String content = msgContent.text();
            log.info("#### content {}",content);
            String name = decodeUTF8(msgContent.text().getBytes());
            log.info("#### name: " + name);
            WebSocketClient webSocketClient = new WebSocketClient(name);
            webSocketClientHashMap.put(ctx.channel().id(),webSocketClient);
        } else if (frame instanceof BinaryWebSocketFrame) {
            BinaryWebSocketFrame msg = (BinaryWebSocketFrame) frame;
            ByteBuf content = msg.content();
            byte[] byteArray = new byte[content.readableBytes()];
            content.readBytes(byteArray);
            if(webSocketClientHashMap.containsKey(ctx.channel().id())){
                webSocketClientHashMap.get(ctx.channel().id()).sendFile(byteArray);
            }
        } else {
        }
    }

    private void saveToFile(byte[] data) {
        try (FileOutputStream fos = new FileOutputStream(FILE_NAME, true)) {
            fos.write(data);
            System.out.println("Data appended to file: " + FILE_NAME);
        } catch (IOException e) {
            e.printStackTrace();
            System.err.println("Error saving data to file: " + e.getMessage());
        }
    }
    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
        super.exceptionCaught(ctx, cause);
        log.error("exception ", cause);
        Channel channel = ctx.channel();
        if(channel.isActive())ctx.close();
    }

    @Override
    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
        log.info("client，channel id" + ctx.channel().id());
    }

}
```

**Comment by chinarui-na** [comment]
<img width="1490" alt="image" src="https://github.com/user-attachments/assets/0a007d41-b720-41cd-8bbb-d974bf996d5c">

**Comment by chinarui-na** [comment]
![48b4f1bd20b09fdf3fc6a88716ea2c4a](https://github.com/user-attachments/assets/ba55ee03-7097-4d60-90f3-b46b485da96e)

**Comment by chinarui-na** [comment]
<img width="1415" alt="image" src="https://github.com/user-attachments/assets/bc453b53-2f57-4662-9245-4642ba5bd37c">

**Comment by amigniter** [comment]
Unfortunately we can't reproduce such an issue, working properly here. Each metadata for caller uuid and the callee is sent properly. Can you please tell me how to reproduce it and of course I will investigate it to see if there is a potential issue.

A potential issue here can be a change that you did before websocket start (printing info). It doesn't work like that, and can cause unpredictable behaviour. When you obtain a session using _switch_core_session_locate_,  the session is under a lock and you must unlock it when you're done with it.

**Comment by amigniter** [comment]
```bash
mm@desktop:~/wsServer$ ./server 
Waiting for incoming connections...

Connection opened, addr: 192.168.0.19
I receive a message: initial1 (8), from: 192.168.0.19, type:1
Connection opened, addr: 192.168.0.19
I receive a message: initial22 (9), from: 192.168.0.19, type:1
Connection closed, addr: 192.168.0.19
Connection closed, addr: 192.168.0.19
```

This a working example, _initial1_ was the initial message from the caller, and _initial22_ from the callee.

**Comment by chinarui-na** [comment]
I delete log. 
I found that `m_initial_meta` would be changed between the start of the websocket connection and the success of the connection, possibly because there was a problem with the source data.

**Comment by amigniter** [comment]
Thanks.
I see. Ok, do some more tests please when you can. As I said, I can't reproduce it and it's working fine here. Can't see how it could possible change. But, if it still happens for you for some reason, I will move it from the class and make it a member of private data of the channel, but first we'll need to find why it's being changed (i see on your pics char pointer points to some weird location).

**Comment by chinarui-na** [comment]
Okay, when I find out, I'll come back here.

**Comment by 2954456878** [comment]
Is there any new progress? I have the same problem.

**Comment by 2954456878** [comment]
![image](https://github.com/user-attachments/assets/d690c3dd-659c-4b14-ac8f-58d8f55eea9a)
metadata value is truncated

**Comment by SumyabazarBaatar** [comment]
i have the same issue. its like a playing coin toss. sometimes i get, sometimes i cant get metadata that i send. if i receive, last parts are truncated sometimes. needs to be fixed.

**Comment by chinarui-na** [comment]
> truncated
You can fix this problem by modifying the code

**Comment by SumyabazarBaatar** [comment]
i dont want to ruin the code. cuz im not good at c++. but in my view, there is like zero problems to send metadata wrong. but i got still strange metadata that i dont want to receive. Example : 

```
2024-08-16 12:02:57.527718 93.83% [DEBUG] mod_audio_stream.c:150 mod_audio_stream cmd: 8e378083-3205-44f6-81e1-18e1de804cab start ws://Address/pub/audiostream mono 16k {"age":1,"type":"sometype","someID":"66becf604af36c1543b13ec8"}
2024-08-16 12:02:57.527718 93.83% [DEBUG] mod_audio_stream.c:81 calling stream_session_init.
2024-08-16 12:02:57.527718 93.83% [DEBUG] audio_streamer_glue.cpp:368 (8e378083-3205-44f6-81e1-18e1de804cab) resampling from 8000 to 16000
2024-08-16 12:02:57.527718 93.83% [DEBUG] audio_streamer_glue.cpp:379 (8e378083-3205-44f6-81e1-18e1de804cab) stream_data_init
2024-08-16 12:02:57.527718 93.83% [DEBUG] mod_audio_stream.c:87 adding bug.
2024-08-16 12:02:57.527718 93.83% [DEBUG] switch_core_media_bug.c:976 Attaching BUG to sofia/external/89801140@202.131.253.177:5060
2024-08-16 12:02:57.527718 93.83% [DEBUG] mod_audio_stream.c:91 setting bug private data.
2024-08-16 12:02:57.527718 93.83% [DEBUG] mod_audio_stream.c:94 exiting start_capture.
2024-08-16 12:02:57.807709 93.83% [DEBUG] audio_streamer_glue.cpp:135 sending initial metadata **sivep**
2024-08-16 12:03:05.867727 93.70% [NOTICE] sofia.c:1065 Hangup sofia/external/><PHONE NUMBER>@IPHost [CS_EXECUTE] [NORMAL_CLEARING]
```

i cant find what kind of code makes my metadata to **sivep**. 
did you find out your problem ? if you do, Can u explain to me why?

**Comment by SumyabazarBaatar** [comment]
i dont want to ruin the code. cuz im not good at c++. but in my view, there is like zero problems to send metadata wrong. but i got still strange metadata that i dont want to receive. Example : 

```
2024-08-16 12:02:57.527718 93.83% [DEBUG] mod_audio_stream.c:150 mod_audio_stream cmd: 8e378083-3205-44f6-81e1-18e1de804cab start ws://Address/pub/audiostream mono 16k {"age":1,"type":"sometype","someID":"66becf604af36c1543b13ec8"}
2024-08-16 12:02:57.527718 93.83% [DEBUG] mod_audio_stream.c:81 calling stream_session_init.
2024-08-16 12:02:57.527718 93.83% [DEBUG] audio_streamer_glue.cpp:368 (8e378083-3205-44f6-81e1-18e1de804cab) resampling from 8000 to 16000
2024-08-16 12:02:57.527718 93.83% [DEBUG] audio_streamer_glue.cpp:379 (8e378083-3205-44f6-81e1-18e1de804cab) stream_data_init
2024-08-16 12:02:57.527718 93.83% [DEBUG] mod_audio_stream.c:87 adding bug.
2024-08-16 12:02:57.527718 93.83% [DEBUG] switch_core_media_bug.c:976 Attaching BUG to sofia/external/<PHONE NUMBER>@IPHost 
2024-08-16 12:02:57.527718 93.83% [DEBUG] mod_audio_stream.c:91 setting bug private data.
2024-08-16 12:02:57.527718 93.83% [DEBUG] mod_audio_stream.c:94 exiting start_capture.
2024-08-16 12:02:57.807709 93.83% [DEBUG] audio_streamer_glue.cpp:135 sending initial metadata sivep
2024-08-16 12:03:05.867727 93.70% [NOTICE] sofia.c:1065 Hangup sofia/external/><PHONE NUMBER>@IPHost [CS_EXECUTE] [NORMAL_CLEARING]
```

i cant find what kind of code makes my metadata to **sivep**. 
did you find out your problem ? if you do, Can u explain the problem to me?

**Comment by amigniter** [comment]
I took a closer look, a few more investigations and I can't reproduce the problem. Tests done on ubuntu 24.04 (FS built from source) and Debian 12 (Version 1.10.12 -release-10222002881-a88d069d6fgit a88d069 2024-08-02 21:02:27Z 64bit installed from repository). In both environments the metadata was not truncated and it was properly sent. The websocket server used is websocket server in C.

> freeswitch@debian> uuid_audio_stream 2ded9fbd-7764-44d9-96ea-f3e445f9f1e4 start ws://192.168.0.19:3001
 mono 8k ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]  
+OK Success

>2024-11-26 13:57:22.606345 99.43% [DEBUG] mod_audio_stream.c:150 mod_audio_stream cmd: 2ded9fbd-7764-44d9-96ea-f3e445f9f1e4 start ws://192.168.0.19:3001 mono 8k ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]
2024-11-26 13:57:22.606345 99.43% [DEBUG] mod_audio_stream.c:81 calling stream_session_init.
2024-11-26 13:57:22.606345 99.43% [DEBUG] audio_streamer_glue.cpp:348 2ded9fbd-7764-44d9-96ea-f3e445f9f1e4: initializing buffer(320) to adjusted 512 bytes
2024-11-26 13:57:22.606345 99.43% [DEBUG] audio_streamer_glue.cpp:376 (2ded9fbd-7764-44d9-96ea-f3e445f9f1e4) no resampling needed for this call
2024-11-26 13:57:22.606345 99.43% [DEBUG] audio_streamer_glue.cpp:379 (2ded9fbd-7764-44d9-96ea-f3e445f9f1e4) stream_data_init
2024-11-26 13:57:22.606345 99.43% [DEBUG] mod_audio_stream.c:87 adding bug.
2024-11-26 13:57:22.606345 99.43% [DEBUG] switch_core_media_bug.c:976 Attaching BUG to sofia/internal/1000@192.168.0.19:5061
2024-11-26 13:57:22.606345 99.43% [DEBUG] mod_audio_stream.c:91 setting bug private data.
2024-11-26 13:57:22.606345 99.43% [DEBUG] mod_audio_stream.c:94 exiting start_capture.
2024-11-26 13:57:22.606345 99.43% [DEBUG] audio_streamer_glue.cpp:135 sending initial metadata ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]

At ws server side I am testing and printing out the message (type 1 is a text message)

>Connection opened, addr: 192.168.0.23
>I receive a message: ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"] (78), from: 192.168.0.23, type:1
>Connection closed, addr: 192.168.0.23

Unfortunately I still can't find any truncation of the initial metadata. Is your FreeSWITCH working properly? Latest tests I did on a fresh debian installation, FS 1.10.12, and JSON array was properly sent.

**Comment by 2954456878** [comment]
Has it been fixed? After pulling the latest update, the problem does not occur

**Comment by amigniter** [comment]
> Has it been fixed? After pulling the latest update, the problem does not occur
Yes. It has been fixed even months ago, in July .. unfortunately, and it was my bad, i created a new branch locally, and all the tests I did were done against that version. I never noticed it has not been pushed to the main branch. Recently when I started addressing this I noticed that, all my tests were good, then i realized it is from a separate branch not even committed. Now it is pushed to the main branch.
